### ISSUE TITLE
Remove SurfaceManager in favor of Substance.Controller

### ISSUE THREAD

**Issue by michael** [substance/substance#63]
Background: I wanted to connect undo/redo to our command API.  However, I'm not sure if the surface scope is always right for this.

``` js
surface.executeCommand('undo');
```

What about document changes that haven't been triggered by an active surface (e.g. an entity reference that has been changed inside a panel in Archivist)? How to undo them?

One idea I had is to move the executeCommand API one level up to the SurfaceManager, which is always available. If there is no active Surface and e.g. `toggleStrong` is executed we could just do nothing. But at least we would always have an interface to execute commands. Also, are there cases where we don't have any surface at all, but use regular UI elements to trigger document changes. Then we would need 
## Substance.Controller

Maybe we should consider introducing a new interface called let's say `Substance.Controller`, replacing the SurfaceManger, but doing more. It would be some kind of higher interface that you can initialize in your app by passing a Substance doc and an app context (used to let the app handle commands if the substance controller did not find a command for it). We could use the Substance Controller to abstract away some infrastructure code, owning:
- Clipboard
- ComponentRegistry
- CommandRegistry
- registered Surfaces
- Keybindings (to be implemented)
- Expose interface for saving a doc (e.g. accessing unsaved changes => isDirty() etc.)

Document Controller would be the bridge between the app (Substance.Component hierarchy) and our more low level interfaces. E.g. a command (which does not have access to the app) could call this.executeCommand('switchState', {contextId: 'editTopicCitation', topicCitationId: 'foo'}); Since there is no command on document controller level it would be passed to the context (the app) which needs to expose an executeCommand interface too and give it a chance to handle it. This would address #61.

Somehow we should try to unify concepts. We have commands now on Surface level, and actions on component level, which are very similar and could be one concept. Also it hurts sometimes that we don't have dependency injection in our Substance (non-component world). But if we wanted that it would mean we need to give up on our plain constructors.

TL;DR: Would a higher level Substance.Controller interface make sense that deals with surfaces, commands like undo/redo, connect the surface editing world with the app world and reduce the app boilerplate?

I could start sketching this as part of next Writer Refactor...

**Comment by obuchtala** [comment]
I think that is reasonable. It sounds like this would be the place to add customizations, too, in future.
Go for it!

**Comment by michael** [comment]
Will do! 

Just one more rant: I think I need to make a distinction between a Command and an Event. 
- **Commands** need a class implementation and they inspect the current editing state (which surface is focussed, what is the selection to perform the appropriate operations. A command either triggers a change to the document or makes the UI somehow change accordingly
- **Events**: I would use the Controller as a global event dispatcher mechanism, where you can emit and and bind/connect to certain events from everywhere, so even leaf components that don't share a parent can communicate with each other.

Two scenarios:
- toggleLink is executed: canExpand is true so the link will be expanded using the expand transformation
- toggleLink again: canEdit is true (because your cursor is under an existing link) so this time no transformations will be triggered but instead we do `controller.emit('show:edit:prompt', linkId)` the linktool listens to that event and shows the link editor.

Good?

**Comment by obuchtala** [comment]
Sounds good. As always with event based systems: don't overuse it ;)

**Comment by michael** [comment]
This is done.
